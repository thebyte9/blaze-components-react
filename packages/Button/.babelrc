{
  "presets": ["@babel/preset-env", "@babel/preset-react", "@babel/preset-typescript"],
  "plugins": [
    [
      "@babel/plugin-transform-runtime",
      {
        "regenerator": true
      }
    ],
    "@babel/plugin-syntax-jsx",
    "@babel/proposal-class-properties",
    "@babel/proposal-object-rest-spread"
  ]
}
