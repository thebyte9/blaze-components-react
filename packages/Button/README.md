
## Description

Custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.

## Usage

* Simple

```js
<Button>Button</Button>
```

* With modifier

```js
<Button modifiers="plain">Plain</Button>

<Button modifiers="small">Small</Button>
```

* Combining modifiers

```js
<Button modifiers="outline rounded">
  Rounded/Outline
</Button>
```

* Disabled

```js
<Button disabled>Alert</Button>
```

* Type submit

```js
<Button submit>Send</Button>
```

## API

##### Buttons can receive a number of `props` as follow:


| NAME   | TYPE | DEFAULT | OPTIONS |
| :---  | :---:  | :---: | ------- |
| modifiers | String | empty | rounded, outline, alert, cta, light, dark, disabled, icon, small, full-width, back, link |
| submit | Boolean | false | 
| disabled | Boolean | false | 
| children | single/array of nodes | null | 
