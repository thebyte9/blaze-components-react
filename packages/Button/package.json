{
  "name": "@blaze-react/button",
  "version": "0.6.0",
  "description": "Custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.",
  "main": "./src/index.ts",
  "scripts": {
    "build": "NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --quiet --fix",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "pre-publish": "yarn build && yarn lint && yarn test"
  },
  "keywords": [],
  "author": "byte9",
  "license": "MIT",
  "peerDependencies": {
    "prop-types": "^15.7.2",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "devDependencies": {
    "webpack": "^5.28.0",
    "webpack-cli": "^4.6.0"
  }
}
