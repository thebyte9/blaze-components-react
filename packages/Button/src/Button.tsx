import React from 'react';
import { buildClassNames } from '@blaze-react/utils';

type TType = 'button' | 'submit' | 'reset';
export interface IButtonProps {
  disabled?: boolean;
  children?: unknown;
  utilities?: string;
  modifiers?: string[];
  type?: TType;
}

export const Button = ({ children, utilities = '', ...rest }: IButtonProps): JSX.Element => {
  const { modifiers = [], disabled, type = 'button' } = rest;

  const DeprecatedButton = () => {
    console.warn('Modifiers will be deprecated in the near future. You should use CSS utilities classes instead');

    const formatedModifiers: string = modifiers.map((modifier) => `button--${modifier}`).join(' ');
    const buttonClassNames: string = buildClassNames('button', {
      [formatedModifiers]: !!modifiers,
    });

    return (
      <button disabled={disabled} className={buttonClassNames} type={type} {...rest}>
        {children}
      </button>
    );
  };

  const BlazeButton = (
    <button type="button" className={utilities} {...rest}>
      {children}
    </button>
  );

  return utilities !== '' ? BlazeButton : DeprecatedButton();
};
