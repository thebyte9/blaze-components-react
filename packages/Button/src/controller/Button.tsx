import React from 'react';
import { DisplayIconProps, IButtonProps } from '../types';
import { ButtonView } from '../view/ButtonView';
import { Icon } from '@blaze-react/icon';
import { useComponentLogic } from '../hooks/useComponentLogic';

export const Button = ({
  label,
  type,
  variant,
  iconOnly = false,
  disabled = false,
  icon,
  displayIcon = DisplayIconProps.NoIcon,
  overrides = [],
  children,
  theme,
  onClick,
}: IButtonProps): JSX.Element => {
  const { display, classes } = useComponentLogic({
    componentVariant: variant,
    theme: theme,
    displayIcon,
    iconOnly: iconOnly,
    overrides: overrides,
  });

  return (
    <ButtonView classes={classes} type={type} disabled={disabled} onClick={onClick}>
      <Icon display={display} iconOnly={iconOnly} label={label} icon={icon} />
      {children}
    </ButtonView>
  );
};
