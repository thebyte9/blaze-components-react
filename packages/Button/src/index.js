/* eslint-disable react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ disabled, submit, children, modifiers, ...attrs }) => {
  const _type = submit ? 'submit' : 'button';

  const _modifiers = modifiers
    .split(' ')
    .map(modifier => `button--${modifier}`)
    .join(' ');

  return (
    <button disabled={disabled} className={`button ${_modifiers}`} type={_type} {...attrs}>
      {children}
    </button>
  );
};

Button.propTypes = {
  modifiers: PropTypes.string,
  disabled: PropTypes.bool,
  submit: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};

Button.defaultProps = {
  modifiers: '',
  disabled: false,
  submit: false,
  children: null
};

export default Button;
