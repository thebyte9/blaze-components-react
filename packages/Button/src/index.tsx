/* eslint-disable react/button-has-type */
import React, { ButtonHTMLAttributes, FunctionComponent } from "react";

interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  disabled?: boolean;
  // submit?: boolean;
  children?: JSX.Element | string;
  modifiers?: string;
  type?: "button" | "submit" | "reset";
}

const Button: FunctionComponent<IButtonProps> = ({
  disabled,
  // submit,
  type,
  children,
  modifiers,
  ...attrs
}): JSX.Element => {
  const allModifiers =
    modifiers &&
    modifiers
      .split(" ")
      .map(modifier => `button--${modifier}`)
      .join(" ");
  return (
    <button
      disabled={disabled}
      className={`button ${allModifiers}`}
      type={type}
      {...attrs}
    >
      {children}
    </button>
  );
};
Button.defaultProps = {
  children: "",
  disabled: false,
  modifiers: ""
};

export default Button;
