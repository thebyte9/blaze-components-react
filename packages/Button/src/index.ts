import { Button } from './view/Button';
import { ButtonController } from './controller/ButtonController';

export { Button, ButtonController };
