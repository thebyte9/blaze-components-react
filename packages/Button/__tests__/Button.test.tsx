import { mount } from "enzyme";
import expect from "expect";
import React from "react";
import Button from "../src/Button";

const testProps = {
  modifiers: ["outline", "rounded"]
};

describe("Button component", () => {
  const wrapper = mount(<Button {...testProps} />);

  test("should be defined and renders correctly (snapshot)", () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".button").length).toBe(1);
  });

  test("should be type button by default", () => {
    expect(wrapper.find(".button").props().type).toBe("button");
  });

  test("should be type submit when passing isSubmit prop", () => {
    expect(mount(<Button type="submit" />).props().type).toBe("submit");
  });
});
