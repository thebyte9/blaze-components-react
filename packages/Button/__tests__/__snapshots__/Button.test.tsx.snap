// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should be defined and renders correctly (snapshot) 1`] = `
<withUtils()
  disabled={false}
  modifiers={
    Array [
      "outline",
      "rounded",
    ]
  }
  type="button"
>
  <Component
    disabled={false}
    modifiers={
      Array [
        "outline",
        "rounded",
      ]
    }
    type="button"
    utils={
      Object {
        "ErrorMessage": [Function],
        "classNames": [Function],
        "parseTextBlock": [Function],
        "removeExtraSpaces": [Function],
        "uniqueId": [Function],
      }
    }
  >
    <button
      className="button button--outline button--rounded"
      disabled={false}
      type="button"
    />
  </Component>
</withUtils()>
`;
